File: CIA - 1/8_b&b_extended_lists.py
import networkx as nx
import matplotlib.pyplot as plt
import heapq

# Branch and Bound with Extended Lists (Dead Horse Principle)
def branch_and_bound_extended(graph, start, goal):
    frontier = [(0, start, [start])]
    visited = set()
    extended_lists = {}

    while frontier:
        cost, node, path = heapq.heappop(frontier)

        if node == goal:
            return path
        
        visited.add(node)
        
        if node not in extended_lists:
            extended_lists[node] = []
        
        for neighbor in graph.neighbors(node):
            if neighbor not in visited and neighbor not in extended_lists[node]:
                edge_cost = graph[node][neighbor]['weight']
                extended_lists[node].append(neighbor)
                heapq.heappush(frontier, (cost + edge_cost, neighbor, path + [neighbor]))

    print("No path found.")
    return None

# Input graph details
num_vertices = int(input("Enter the number of vertices: "))
edges = []
for _ in range(int(input("Enter the number of edges: "))):
    edge = input("Enter edge (format: source target weight): ").split()
    edges.append((edge[0], edge[1], int(edge[2])))

start_vertex = input("Enter the start vertex: ")
goal_vertex = input("Enter the goal vertex: ")

# Create a weighted graph using networkx
G = nx.Graph()
for edge in edges:
    G.add_edge(edge[0], edge[1], weight=edge[2])

# Perform Branch and Bound with Extended Lists
path = branch_and_bound_extended(G, start_vertex, goal_vertex)

# Draw the initial graph
pos = nx.spring_layout(G)
plt.figure(figsize=(10, 7))
nx.draw(G, pos, with_labels=True, node_size=700, node_color='lightblue', font_size=16, font_weight='bold', edge_color='gray')
edge_labels = nx.get_edge_attributes(G, 'weight')
nx.draw_networkx_edge_labels(G, pos, edge_labels=edge_labels)
plt.title("Initial Graph")
plt.show()

# If a path is found, display the path
if path:
    path_edges = list(zip(path, path[1:]))
    plt.figure(figsize=(10, 7))
    nx.draw(G, pos, with_labels=True, node_size=700, node_color='lightblue', font_size=16, font_weight='bold', edge_color='gray')
    nx.draw_networkx_edges(G, pos, edgelist=path_edges, edge_color='r', width=2)
    nx.draw_networkx_nodes(G, pos, nodelist=path, node_color='lightgreen')
    plt.title(f"Branch and Bound with Extended Lists Path: {path}")
    plt.show()
else:
    print("No path found.")

File: CIA - 1/8_b&b_extended_lists.c
#include <stdio.h>
#include <limits.h>
#include <stdbool.h>
#include <string.h>

#define MAX 100

int adjMatrix[MAX][MAX], numVertices;
char vertices[MAX];
int bestCost = INT_MAX;
bool visited[MAX];
bool extendedList[MAX];
int currentPath[MAX];

// Function to display the graph as adjacency matrix
void displayGraph() {
    printf("\nAdjacency Matrix:\n");
    for (int i = 0; i < numVertices; i++) {
        printf("%c: ", vertices[i]);
        for (int j = 0; j < numVertices; j++) {
            printf("%d ", adjMatrix[i][j]);
        }
        printf("\n");
    }
}

// Recursive Branch and Bound function with extended lists
void branchAndBoundExtended(int current, int goal, int cost, int depth) {
    if (cost >= bestCost) {
        return; // Prune branch if the current cost exceeds the best found
    }

    if (current == goal) {
        if (cost < bestCost) {
            bestCost = cost;
            printf("New best path with cost %d: ", bestCost);
            for (int i = 0; i < depth; i++) {
                printf("%c -> ", vertices[currentPath[i]]);
            }
            printf("%c\n", vertices[goal]);
        }
        return;
    }

    for (int i = 0; i < numVertices; i++) {
        if (adjMatrix[current][i] != 0 && !visited[i] && !extendedList[i]) {
            visited[i] = true;
            extendedList[i] = true; // Mark in extended list to avoid revisiting
            currentPath[depth] = i;
            branchAndBoundExtended(i, goal, cost + adjMatrix[current][i], depth + 1);
            visited[i] = false;
        }
    }
}

// Main function
int main() {
    int i, j, source, destination;
    char start, goal;

    printf("Enter the number of vertices: ");
    scanf("%d", &numVertices);

    printf("Enter the names of the vertices:\n");
    for (i = 0; i < numVertices; i++) {
        scanf(" %c", &vertices[i]);
    }

    printf("Enter the number of edges: ");
    int numEdges;
    scanf("%d", &numEdges);

    // Initialize adjacency matrix
    for (i = 0; i < numVertices; i++) {
        for (j = 0; j < numVertices; j++) {
            adjMatrix[i][j] = 0;
        }
    }

    printf("Enter the edges (source destination weight):\n");
    for (i = 0; i < numEdges; i++) {
        char s, d;
        int weight;
        scanf(" %c %c %d", &s, &d, &weight);
        source = strchr(vertices, s) - vertices;
        destination = strchr(vertices, d) - vertices;
        adjMatrix[source][destination] = weight;
        adjMatrix[destination][source] = weight; // Assuming undirected graph
    }

    displayGraph();

    // Taking start and goal inputs
    printf("Enter the start vertex: ");
    scanf(" %c", &start);
    printf("Enter the goal vertex: ");
    scanf(" %c", &goal);

    source = strchr(vertices, start) - vertices;
    destination = strchr(vertices, goal) - vertices;

    // Initialize visited and extended lists
    for (i = 0; i < numVertices; i++) {
        visited[i] = false;
        extendedList[i] = false;
    }

    visited[source] = true;
    extendedList[source] = true;
    currentPath[0] = source;

    // Perform Branch and Bound Search with extended list (Dead Horse Principle)
    branchAndBoundExtended(source, destination, 0, 1);

    printf("Best cost found: %d\n", bestCost);

    return 0;
}
